### PR TITLE
[SourceKit] Add subscript to doc structure (SR-5035)

### PR DESCRIPTION
**Pull request by marcelofabri**
This adds `subscript` to the structure reported by SourceKit.

For example, for this file:
```swift
class Foo {
    subscript(index: Int) -> Int {
        get {
            return 0
        }
        set {
            print(newValue)
        }
    }
}
```

We get this structure now:
```
{
  key.offset: 0,
  key.length: 104,
  key.diagnostic_stage: source.diagnostic.stage.swift.parse,
  key.substructure: [
    {
      key.kind: source.lang.swift.decl.class,
      key.accessibility: source.lang.swift.accessibility.internal,
      key.name: "Foo",
      key.offset: 0,
      key.length: 103,
      key.runtime_name: "_TtC8__main__3Foo",
      key.nameoffset: 6,
      key.namelength: 3,
      key.bodyoffset: 11,
      key.bodylength: 91,
      key.substructure: [
        {
          key.kind: source.lang.swift.decl.function.subscript,
          key.accessibility: source.lang.swift.accessibility.internal,
          key.setter_accessibility: source.lang.swift.accessibility.internal,
          key.name: "subscript(_:)",
          key.offset: 13,
          key.length: 88,
          key.nameoffset: 0,
          key.namelength: 0,
          key.bodyoffset: 43,
          key.bodylength: 57,
          key.substructure: [
            {
              key.kind: source.lang.swift.decl.var.parameter,
              key.name: "index",
              key.offset: 23,
              key.length: 10,
              key.typename: "Int",
              key.nameoffset: 0,
              key.namelength: 0
            }
          ]
        },
        {
          key.kind: source.lang.swift.expr.call,
          key.name: "print",
          key.offset: 79,
          key.length: 15,
          key.nameoffset: 79,
          key.namelength: 5,
          key.bodyoffset: 85,
          key.bodylength: 8
        }
      ]
    }
  ]
}
```

Resolves [SR-5035](https://bugs.swift.org/browse/SR-5035).

This wouldn't be possible without the help from @johnfairh and @CodaFi. Thanks so much! 🙌

### CODE REVIEW

File: include/swift/IDE/SyntaxModel.h [hunk 1]
```diff
@@ -146,6 +147,7 @@ struct SyntaxStructureNode {
     case SyntaxStructureKind::StaticVariable:
     case SyntaxStructureKind::ClassVariable:
     case SyntaxStructureKind::Parameter:
+    case SyntaxStructureKind::Subscript:
```
Comment (marcelofabri) [inline include/swift/IDE/SyntaxModel.h]: I'm not sure if this is the proper solution or if it's better to:
1. Rename this function
2. Whenever using `isVariable` also check if `Kind == Subscript`

**Comment by marcelofabri** [comment]
cc @nkcsgexi

File: include/swift/IDE/SyntaxModel.h (hunk 1 above)
Comment (nkcsgexi) [inline include/swift/IDE/SyntaxModel.h]: what happens if we don't add this line here?

Comment (marcelofabri) [inline include/swift/IDE/SyntaxModel.h]: the syntax kinds of `get` and `set` inside `subscript`s become `identifier` instead of `keyword`

Comment (CodaFi) [inline include/swift/IDE/SyntaxModel.h]: This accessor (`isVariable()`) is only being used in one place in the Syntax Model and only for things that have substructure that needs to be walked.  Would it be OK to just rename it `hasSubstructure()` or something?

Comment (nkcsgexi) [inline include/swift/IDE/SyntaxModel.h]: I agree with @CodaFi . Renaming it sounds a right approach to me.

Comment (marcelofabri) [inline include/swift/IDE/SyntaxModel.h]: Renamed and amended the commit.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Pull request merged**